vembrane/backend/base.py

```diff
@@ -0,0 +1,272 @@
+from abc import abstractmethod, abstractproperty
+from enum import Enum
+from typing import List, Optional, Tuple
+
+from ..ann_types import NA
+from ..errors import UnknownSample
+
+
+class Backend(Enum):
+    pysam = (0,)
+    cyvcf2 = 1
+
+    def __str__(self):
+        return self.name.lower()
+
+    def __repr__(self):
+        return str(self)
+
+    @staticmethod
+    def from_string(s):
+        try:
+            return Backend[s]
+        except KeyError:
+            return s
+
+
+class VCFRecordInfo:
+    @abstractmethod
+    def __contains__(self, item):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(self, item):
+        raise NotImplementedError
+
+    @abstractmethod
+    def __setitem__(self, key, value):
+        raise NotImplementedError
+
+    def get(self, key, default=None):
+        if key in self:
+            return self[key]
+        return default
+
+
+class VCFRecordSamples:
+    pass
+
+
+class NoValueDict:
+    def __contains__(self, item):
+        try:
+            value = self[item]
+        except KeyError:
+            return False
+        return value is not NA

```

Comment:
So the consense was `false` and Sven a warning, which itself is not implemented. Do we want to have such a warning for each missing value in each read?

Reply:
If a VCF file has `some_random_key` specified in its header, every record should have that field set, which is why the VCF spec has that special value MISSING (i.e. `.`).
If one of a million records for some reason does not have `some_random_key` specified, I'd just take it to mean `some_random_key=.`, which means that the information is missing for that key.

I have added a warning printing both the record id and the record itself in that case; that might be annoying for files which have a lot of records which simply omit info keys, instead of using `.` as corresponding values.

I have tried finding information about this in the VCF spec, but couldn't find anything definitive as to what is the preferred way. `bcftools` at least does not complain about this case.